src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -343,64 +345,35 @@ generateX x
              $ hoistEither $ errorNoSuggestions (ErrorFunctionWrongArgs a x fErr [])
            annotate Map.empty resT $ \a' -> Prim a' p
 
-    -- Cases require all their alternatives to have the same type, as well as the scrutinee and the alternatives to have
-    -- the same temporality.
-    -- This may be more restrictive than necessary, since lets allow different temporalities, but hopefully will not be an issue.
-    --
-    -- TODO: cases with nested constructors must be converted to multiple nested cases.
-    -- For example:
-    -- > case x
-    -- >  | Some (Some v, None) -> v
-    -- >  | Some (None, Some v) -> v
-    -- >  | _                   -> 0
-    -- > end
-    --
-    -- this should be flattened to
-    -- > case x
-    -- > | Some y
-    -- > -> case y
-    -- >    | (l,r)
-    -- >    -> case l
-    -- >       | Some v
-    -- >       -> case r
-    -- >          | Some _
-    -- >          -> 0
-    -- >          | None
-    -- >          -> v
-    -- >          end
-    -- >       | None
-    -- >       -> case r
-    -- >          | Some v
-    -- >          -> v
-    -- >          | None
-    -- >          -> 0
-    -- >          end
-    -- >       end
-    -- >    end
-    -- > | None
-    -- > -> 0
+    -- Cases require:
+    --  * alternatives to have "join-able" types, i.e. no mixing of aggregates and elements.
     --
     Case a scrut pats
      -> do (scrut', sub) <- generateX scrut
            (pats', subs) <- unzip <$> mapM (generateP $ annResult $ annotOfExp scrut') pats
 
-           resT <- case pats' of
-                    [] -> hoistEither
-                        $ errorNoSuggestions
-                        $ ErrorEmptyCase a x
-                    ((_,alt):_)
-                       -> return $ annResult $ annotOfExp alt
+           let scrutT  =  annResult $ annotOfExp scrut'
+           resT        <- case pats' of
+                           [] -> hoistEither
+                               $ errorNoSuggestions
+                               $ ErrorEmptyCase a x
+                           ((_,alt):_)
+                              -> return $ annResult $ annotOfExp alt
```

Comment:
yeah actually I think you need a fresh variable for each alternative. and you start by joining the scrut and the first alternative, put that in a fresh var. then join the last fresh and the next alternative, and put it in a new fresh. then at the end the last fresh will be the join of all the scrut and alternatives.

wow! that's weird! maybe we can actually change the Join constraint to be (Join (Type n) [Type n]) so the output is a single var but takes multiple inputs